packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -3687,7 +3687,8 @@ abstract class Element extends DiagnosticableTree implements BuildContext {
 
   // The children that have been forgotten by forgetChild. This will be used in
   // [update] to remove the global key reservations of forgotten children.
-  final Set<Element> _debugForgottenChildrenWithGlobalKey = HashSet<Element>();
+  // Initialze to null in profile/release mode to avoid allocating unused sets
+  final Set<Element>? _debugForgottenChildrenWithGlobalKey = kDebugMode ? HashSet<Element>() : null;
```

Comment:
Why isn't this tree shaken away as a write only field?

Reply:
🤷 is it write only?

Reply:
Yes - it's assigned on creation and then never used again outside of asserts. @alexmarkov do you know why this would still show up in profile builds using --aot --tfa?

Reply:
Do those flags apply in profile mode? (I think so)

Reply:
Are you sure it's this one here and not the one in `_updateInheritence`?

Looking at a release snapshot this is definitely shaken away. It looks like it is in the profile one as well if I'm reading the tooling output correctly.

Reply:
Definitely not, will take another pass a this tonight

Reply:
@dnfield I built Flutter Gallery in profile mode, dumped app.dill and field `_debugForgottenChildrenWithGlobalKey` is not there. Where do you see it?

On a side note, I think it may be useful to cut down the field initializer, as initializer is still executed (in constructor) even if field is removed, as initializer may have arbitrary side effects.

Reply:
Maybe what I'm seeing in the profile is the initializer then?

Reply:
if I'm understanding this correctly - the field being removed but the initializer still running would lead to allocating and then eventually releasing a HashSet per element?

Reply:
Ok. that makes sense. It does have side effects! We just don't like them.

Reply:
If we made this `late final` would the initializer still be run?

Reply:
I'm not sure if it ever happened but I recall discussions about making late final run eagerly as an optimization. I don't think the initialization is defined to be lazy the same way toplevel/statics are - that would make it a bit fragile IMO.

Reply:
`late final` makes it lazy. And the following app works to not run the initialization:

```dart
void main() {
  final Foo foo = Foo();
  assert(foo._debugBar != null);
}

class Foo {
  late final Bar _debugBar = Bar();
}

class Bar {
  Bar() {
    print('Initializing a Bar!');
  }
}
```

Reply:
Initializers of `late final` fields don't run in constructors - they run lazily when getter is called. So, yes, in this case `late final` may be useful to avoid overhead of running initializer. Note that write-only `late final` fields _without_ initializer are not going to be tree shaken as they needed to guard against repetitive writes.

Reply:
https://dart.dev/null-safety/understanding-null-safety#lazy-initialization says this is the way it's supposed to work.

Reply:
seems like a good deal!

Reply:
Digging a little more on this, `late final` does not end up letting the field get tree shaken, but your version here does.

Reply:
Oh, I think we'd probably want to eliminate both the field and initializer. WDUT about reopening this (and applying it to all the complex fields ?)

Reply:
Sgtm